src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -202,21 +216,116 @@ public boolean isOpen()
       public void close() throws IOException
       {
         open = false;
+        useDelegateSmooshWriter = false;
         internalFiles.put(name, new Metadata(currOut.getFileNum(), startOffset, currOut.getCurrOffset()));
 
         if (bytesWritten != currOut.getCurrOffset() - startOffset) {
-          throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
-        }
-        if (bytesWritten != size) {
-          throw new IOException(
-              String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
-          );
+            throw new ISE("WTF? Perhaps there is some concurrent modification going on?");
+          }
+          if (bytesWritten != size) {
+            throw new IOException(
+                String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
+            );
+          }        
+        if(!mergeOnProgress && !useDelegateSmooshWriter)
+        {
+        	mergeWithSmoosher();
         }
+
       }
     };
   }
+  
+  public OutputStream makeOutputStream(String filename) throws IOException
+  {
+    File retFile = createdFiles.get(filename);
+    if (retFile == null) {
+      retFile = File.createTempFile("smoosher", filename);
+      retFile.deleteOnExit();
+      createdFiles.put(filename, retFile);
+      return new BufferedOutputStream(new FileOutputStream(retFile));
+    } else  {
+      return new BufferedOutputStream(new FileOutputStream(retFile));
+    } 
+  }
+  
+  public InputStream makeInputStream(String filename) throws IOException
+  {
+    final File retFile = createdFiles.get(filename);
+    return retFile == null ? null : new FileInputStream(retFile);
+  }
+  
+  private synchronized void mergeWithSmoosher() throws IOException
+  {
+	  //get processed elements from the stack and write.
+	  mergeOnProgress = true;
+	 
+	  while(! files.isEmpty())
+	  {
+		  String file = files.pop();
+		  add(file, createdFiles.get(file));
+		  createdFiles.get(file).delete();
+		  createdFiles.remove(file);
+	  }
+	  mergeOnProgress = false;
+  }
+  
+  private SmooshedWriter delegateSmooshedWriter(final String name,final long size) throws IOException
+  {
+		
+		return new SmooshedWriter () {
```

Comment:
There are now two implementations of `SmooshedWriter` in here and for some reason, they don't do the same sanity checks on values.  Why?